src/track/globaltrackcache.cpp

```diff
@@ -576,98 +630,118 @@ TrackRef GlobalTrackCache::initTrackId(
     return trackRefWithId;
 }
 
-void GlobalTrackCache::evictOrDelete(
+void GlobalTrackCache::evictAndSave(
         Track* plainPtr) {
     DEBUG_ASSERT(plainPtr);
+
+    // Nearly everything is possible until the cache is locked!!!
     GlobalTrackCacheLocker cacheLocker;
 
-    const AllocatedTracks::iterator allocatedTrack =
-            m_allocatedTracks.find(plainPtr);
-    if (allocatedTrack == m_allocatedTracks.end()) {
-        // We have handed out and already delete this track while waiting at
+    // After obtaining the lock we need to check if the deletion
+    // is still valid, i.e. if the pointer has not already been
+    // deleted or if it has been revived by an intermediate cache
+    // lookup while this thread was blocked.
+    const CachedTracks::iterator cachedTrack =
+            m_cachedTracks.find(plainPtr);
+    if (cachedTrack == m_cachedTracks.end()) {
+        // We have already deleted this track while waiting at
         // the lock at the beginning of this function.
         if (debugLogEnabled()) {
             kLogger.debug()
-                    << "Skip deletion of already deleted track"
+                    << "Skip to evict and save an already evicted track"
                     << plainPtr;
         }
         return;
     }
-
-    if (!allocatedTrack->second.expired()) {
-        // We have handed out this track again while waiting at
-        // the lock at the beginning of this function.
+    if (!cachedTrack->second.expired()) {
+        // We have handed out (revived) this track again while waiting
+        // at the lock at the beginning of this function or a new track
+        // object has been allocated with the same memory address.
         if (debugLogEnabled()) {
             kLogger.debug()
-                    << "Skip deletion of revived track"
+                    << "Skip to evict and save a revived or reallocated track"
                     << plainPtr;
         }
         return;
     }
 
-    TrackPointer strongPtr;
-
-    strongPtr = lookupById(plainPtr->getId());
-    if (!strongPtr) {
-        const auto trackRef = createTrackRef(*plainPtr);
-        strongPtr = lookupByRef(trackRef);
-    }
-
-    if (strongPtr) {
-        evictAndSave(strongPtr);
-    } else {
-        // Track has already been evicted and can be deallocated now.
-        m_allocatedTracks.erase(allocatedTrack);
```

Comment:
Here is the place where the last reference to the reference counting object falls to zero and is deleted. 
This is quite save, because the expired check still works on the second run of the deleter.  
This is an extra safety net when deactivate(). In the solution from this PR we have no guarantee that we do not end up in two independent SharedPointer with a deleter.